extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp

```diff
@@ -0,0 +1,310 @@
+  }
+};
+
+class SimpleSSLTestClientTLSv1_2  : public SimpleSSLTestClient {
+ public:
+  SimpleSSLTestClientTLSv1_2(const std::string& host, const std::string& port)
+      : SimpleSSLTestClient(TLSv1_2_client_method(), host, port) {
+  }
+};
+
+class TLSServerSocketSupportedProtocolsTest {
+ public:
+    explicit TLSServerSocketSupportedProtocolsTest(const std::string& key_dir)
+        : is_running_(false), key_dir_(key_dir), configuration_(std::make_shared<minifi::Configure>()) {
+    }
+
+    void run() {
+      configureSecurity();
+
+      createServerSocket();
+
+      verifyTLSServerSocketExclusiveCompatibilityWithTLSv1_2();
+
+      shutdownServerSocket();
+    }
+
+ protected:
+    void configureSecurity() {
+      host_ = org::apache::nifi::minifi::io::Socket::getMyHostName();
+      port_ = "38778";
+      if (!key_dir_.empty()) {
+        configuration_->set(minifi::Configure::nifi_remote_input_secure, "true");
+        configuration_->set(minifi::Configure::nifi_security_client_certificate, key_dir_ + "cn.crt.pem");
+        configuration_->set(minifi::Configure::nifi_security_client_private_key, key_dir_ + "cn.ckey.pem");
+        configuration_->set(minifi::Configure::nifi_security_client_pass_phrase, key_dir_ + "cn.pass");
+        configuration_->set(minifi::Configure::nifi_security_client_ca_certificate, key_dir_ + "nifi-cert.pem");
+        configuration_->set(minifi::Configure::nifi_default_directory, key_dir_);
+      }
+    }
+
+    void createServerSocket() {
+      const auto socket_context = std::make_shared<org::apache::nifi::minifi::io::TLSContext>(configuration_);
+      server_socket_ = std::make_shared<org::apache::nifi::minifi::io::TLSServerSocket>(socket_context, host_, std::stoi(port_), 3);
+      assert(0 == server_socket_->initialize());
+
+      is_running_ = true;
+      check_ = [this]() -> bool {
+        return is_running_;
+      };
+      handler_ = [this](std::vector<uint8_t> *bytes_written, int *size) {
+        std::string contents = "hello world";
+        *bytes_written = {contents.begin(), contents.end()};
+        bytes_written->push_back(0);
+        *size = bytes_written->size();
+        return *size;
+      };
+      server_socket_->registerCallback(check_, handler_, std::chrono::milliseconds(50));
+    }
+
+    void verifyTLSServerSocketExclusiveCompatibilityWithTLSv1_2() {
+      verifyTLSProtocolIncompatibility<SimpleSSLTestClientTLSv1>();
+      verifyTLSProtocolIncompatibility<SimpleSSLTestClientTLSv1_1>();
+      verifyTLSProtocolCompatibility<SimpleSSLTestClientTLSv1_2>();
+    }
+
+    template <class TLSTestClient>
+    void verifyTLSProtocolIncompatibility() {
+      verifyTLSProtocolCompatibility<TLSTestClient>(false);
+    }
+
+    template <class TLSTestClient>
+    void verifyTLSProtocolCompatibility() {
+      verifyTLSProtocolCompatibility<TLSTestClient>(true);
+    }
+
+    template <class TLSTestClient>
+    void verifyTLSProtocolCompatibility(bool should_be_compatible) {
```

Comment:
I would argue that the templates forwarding to this template are unnecessary and we could call this one directly from `verifyTLSServerSocketExclusiveCompatibilityWithTLSv1_2 `

Reply:
fixed